### PR TITLE
[Core] Add geometry python interfaces

### PR DESCRIPTION
**Pull request by tteschemacher**
**Description**
This PR adds a couple of python interfaces to the geometry. It also removes interfaces which would run into errors (GetFastNormal and GetFastUnitNormal -  here the local location is always created on 0,0 which doesn't make sense)

**Changelog**
Python interfaces are added for:
- IntegrationPointsNumber
- Normal with IntegrationPointIndex
- UnitNormal with IntegrationPointIndex
- Jacobian
- DeterminantOfJacobian
- ShapeFunctionsValues
- ShapeFunctionDerivatives

### CODE REVIEW

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Please, replace all these methods with lambdas

**Comment by tteschemacher** [comment]
> 
> 
> Please, replace all these methods with lambdas

Good that I've got nothing better to do than converting things into lambdas ;) it's done.

**Review by loumalouomega — DISMISSED** [review]

**Comment by tteschemacher** [comment]
thanks @loumalouomega for your review and comment! :)

**Review by philbucher — DISMISSED** [review]
:+1:

**Comment by tteschemacher** [comment]
@loumalouomega the tests are failing in the test_exact_integration because I removed the corrupted interface. Could you check what is wrong there?

**Comment by loumalouomega** [comment]
> @loumalouomega the tests are failing in the test_exact_integration because I removed the corrupted interface. Could you check what is wrong there?

Later, right now I cannot

**Comment by loumalouomega** [comment]
> > @loumalouomega the tests are failing in the test_exact_integration because I removed the corrupted interface. Could you check what is wrong there?
> 
> Later, right now I cannot

Please do not remove methods, it may be used by someone else, like me

**Review by loumalouomega — DISMISSED** [review]

**Comment by tteschemacher** [comment]
If they're used by someone it would be better for them to know it. Because they really don't produce anything good :P

**Comment by philbucher** [comment]
how about adding deprecation warnings to those functions?

**Review by philbucher — DISMISSED** [review]
please consider adding a deprecation warning to the critical function

**Comment by loumalouomega** [comment]
> please consider adding a deprecation warning to the critical function

What I don't understand is why do we need to deprecate those methods

**Comment by philbucher** [comment]
> > please consider adding a deprecation warning to the critical function
> 
> What I don't understand is why do we need to deprecate those methods

I think because it gives the normal at an undefined location but @tteschemacher knows better

**Comment by loumalouomega** [comment]
> > > please consider adding a deprecation warning to the critical function
> > 
> > 
> > What I don't understand is why do we need to deprecate those methods
> 
> I think because it gives the normal at an undefined location but @tteschemacher knows better

Lines and triangles (which are 95% of conditions real cases) only have one integration point, and for the quadrilateral (the other 4.8%, the 0.2 lets says corresponds with the quadratic entities) the zero local coordinates correspond with the center. This is very practical because you don't need to use dummy local coordinates, that many times are not trivial

**Comment by loumalouomega** [comment]
> > > > please consider adding a deprecation warning to the critical function
> > > 
> > > 
> > > What I don't understand is why do we need to deprecate those methods
> > 
> > 
> > I think because it gives the normal at an undefined location but @tteschemacher knows better
> 
> Lines and triangles (which are 95% of conditions real cases) only have one integration point, and for the quadrilateral (the other 4.8%, the 0.2 lets says corresponds with the quadratic entities) the zero local coordinates correspond with the center. This is very practical because you don't need to use dummy local coordinates, that many times are not trivial

[Here an explanation of the numbers I used](https://www.youtube.com/watch?v=7tzfl1wTemM)

**Comment by philbucher** [comment]
hm sounds reasonable, but this also means that using it for quads is wrong no?

**Comment by loumalouomega** [comment]
> hm sounds reasonable, but this also means that using it for quads is wrong no?

If you want to get the normal in the center (usually you do), it is not

**Comment by loumalouomega** [comment]
> > hm sounds reasonable, but this also means that using it for quads is wrong no?
> 
> If you want to get the normal in the center (usually you do), it is not

A debug message can be throw in other cases...

**Comment by philbucher** [comment]
maybe calling this one `NormalInCenter` then? Or sth similar

**Comment by loumalouomega** [comment]
> maybe calling this one `NormalInCenter` then? Or sth similar

It is only the center for geometries where zero local coordinates are the center (triangles is the corner)

**Comment by philbucher** [comment]
ok for me then

if the method stays then please add some comment explaining this!

**Review by loumalouomega — DISMISSED** [review]

**Review by loumalouomega — DISMISSED** [review]

File: kratos/python/add_geometries_to_python.cpp [hunk 1]
```diff
@@ -153,22 +115,50 @@ void  AddGeometriesToPython(pybind11::module& m)
...
+    .def("Normal", [](GeometryType& self, CoordinatesArrayType& LocalCoords)
+        { return(self.Normal(LocalCoords)); })
+    .def("Normal", [](GeometryType& self, IndexType IntegrationPointIndex)
+        { return(self.Normal(IntegrationPointIndex)); })
+    .def("UnitNormal", [](GeometryType& self)
+        { const auto& r_type = self.GetGeometryType(); KRATOS_WARNING_IF("Geometry", !(r_type == GeometryData::KratosGeometryType::Kratos_Line2D2 || r_type == GeometryData::KratosGeometryType::Kratos_Triangle3D3)) << "WARNING:: Your geometry is not linear, please provide local coordinates or a GP index. Normal will be computed in zero local coordinates" << std::endl; CoordinatesArrayType LocalCoords; LocalCoords.clear(); return(self.UnitNormal(LocalCoords)); })
```
Comment (philbucher) [inline kratos/python/add_geometries_to_python.cpp]: this is where lambdas are not really that useful any more, one can barely read what is happening ;)

```suggestion
        { const auto& r_type = self.GetGeometryType(); KRATOS_WARNING_IF("Geometry", !(r_type == GeometryData::KratosGeometryType::Kratos_Line2D2 || r_type == GeometryData::KratosGeometryType::Kratos_Triangle3D3)) << "Your geometry is not linear, please provide local coordinates or a GP index. Normal will be computed in zero local coordinates" << std::endl; CoordinatesArrayType LocalCoords; LocalCoords.clear(); return(self.UnitNormal(LocalCoords)); })
```

**Review by philbucher — APPROVED** [review]

**Comment by tteschemacher** [comment]
Thanks a lot to everyone for the help and the review!! 👍 :)

**Pull request merged**